File: examples/buyer-node/buyer.mjs
import { wrapFetchWithPaymentFromConfig } from "@x402/fetch";
import { ExactEvmScheme } from "@x402/evm/exact/client";
import { privateKeyToAccount } from "viem/accounts";

const privateKey = process.env.ONECENT_BUYER_PRIVATE_KEY;
if (!privateKey) throw new Error("ONECENT_BUYER_PRIVATE_KEY is required");

const MAX_ATOMIC_USDC = 1000n; // Hard cap: 0.001 USDC.
const signer = privateKeyToAccount(privateKey);
const paidFetch = wrapFetchWithPaymentFromConfig(fetch, {
  schemes: [{ network: "eip155:8453", client: new ExactEvmScheme(signer) }],
  paymentRequirementsSelector: (_version, accepts) => {
    const exactBase = accepts.filter(
      (item) =>
        item.scheme === "exact" &&
        item.network === "eip155:8453" &&
        BigInt(item.amount) <= MAX_ATOMIC_USDC,
    );
    if (exactBase.length !== 1) throw new Error("No single safe payment option under cap");
    return exactBase[0];
  },
});

const response = await paidFetch("https://1cent.maxzoa.ru/v1/url/status", {
  method: "POST",
  headers: { "content-type": "application/json" },
  body: JSON.stringify({ url: "https://example.com/", fresh: false }),
});
console.log(response.status, response.headers.get("X-Request-ID"));
console.log(await response.json());

File: examples/buyer-python/buyer.py
"""One explicitly capped x402 call. Review `onecent doctor` first."""

import asyncio
import os

from eth_account import Account
from x402 import x402Client
from x402.client import max_amount
from x402.http.clients import x402HttpxClient
from x402.mechanisms.evm import EthAccountSigner
from x402.mechanisms.evm.exact.register import register_exact_evm_client

MAX_ATOMIC_USDC = 1_000  # Hard cap: 0.001 USDC.


async def main() -> None:
    account = Account.from_key(os.environ["ONECENT_BUYER_PRIVATE_KEY"])
    client = x402Client().register_policy(max_amount(MAX_ATOMIC_USDC))
    register_exact_evm_client(client, EthAccountSigner(account))
    async with x402HttpxClient(client, timeout=30, follow_redirects=False) as http:
        response = await http.post(
            "https://1cent.maxzoa.ru/v1/url/status",
            json={"url": "https://example.com/", "fresh": False},
        )
        await response.aread()
    print(response.status_code, response.headers.get("X-Request-ID"))
    print(response.json())


if __name__ == "__main__":
    asyncio.run(main())
